### PR TITLE
[gh-1552] refactor encoding and decoding, move to moveos_stdlib.

### LINKED ISSUE #1552: [crypto] refactor & add crypto & bitcoin libraries.
There are several enhancements which can be improved with crypto libraries as mentioned in the April 11 meeting:

1. Split general and bitcoin-specific address encoding & decoding, including base58 and bech32.
2. Should ripemd160 function be exported.

### PR DESCRIPTION
**Pull request by feliciss**
## Summary

Follow the fix described on thread: https://discord.com/channels/1078938449974935592/1227995346731274370
- Closes #1552

### CODE REVIEW

File: frameworks/moveos-stdlib/src/natives/moveos_stdlib/base58.rs [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+
+    let mut cost = gas_params.base;
+
+    let encoded_address_bytes = pop_arg!(args, VectorRef);
+
+    let input_bytes = encoded_address_bytes.as_bytes_ref().to_vec();
+    cost += gas_params.per_byte * NumBytes::new(input_bytes.len() as u64);
+
+    let Ok(bs58_raw_bytes) = bs58::decode(input_bytes).into_vec() else {
+        return Ok(NativeResult::err(cost, E_DECODE_FAILED));
```
Comment (jolestar) [inline frameworks/moveos-stdlib/src/natives/moveos_stdlib/base58.rs:104]: Also define the E_DECODE_FAILED error code in base58.move

File: frameworks/moveos-stdlib/src/natives/moveos_stdlib/bech32.rs [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+        + (gas_params.per_byte * NumBytes::new(public_key.as_bytes_ref().len() as u64));
+
+    let (hrp, variant) = if witness_version == 0 {
+        ("bech32", Variant::Bech32)
+    } else {
+        ("bech32m", Variant::Bech32m)
+    };
+
+    let encoded =
+        bech32::encode(hrp, public_key.as_bytes_ref().to_vec().to_base32(), variant).unwrap();
```
Comment (jolestar) [inline frameworks/moveos-stdlib/src/natives/moveos_stdlib/bech32.rs:49]: Need to handle the error.

**Comment by jolestar** [comment]
Remove the toolchain change and merge this PR first?

**Comment by feliciss** [comment]
> Remove the toolchain change and merge this PR first?

Fixed.

File: frameworks/moveos-stdlib/src/natives/moveos_stdlib/bech32.rs (hunk 2 above)
Comment (feliciss) [reply frameworks/moveos-stdlib/src/natives/moveos_stdlib/bech32.rs:50]: Will handle it in another PR.

**Review by jolestar — APPROVED** [review]

**Pull request merged**